### DISCUSSION TITLE
prisma-client-js not generating the optionals fileds for typescript project.

### DISCUSSION THREAD

**Discussion by jega-ms** [prisma/prisma#18360] [Q&A] [answered]
### Bug description

i'm trying to generate the schema for UserEntity class for typescript project. but it is generating the plain model entity class with null  and ignored the optional  fields ("?").  e.g mobile and emailVerified 

### How to reproduce

generate the schema using this command "npx prisma generate" 

### Expected behavior

Generator should generte this class 

```
export type UserEntity = {
  id: number
  email: string
  firstName: string
  lastName: string
  emailVerified?: boolean| null
  mobile?: string | null
}
```

### Prisma information

Prisma Client (4.11.0 | library) to ./node_modules/@prisma/client in 84ms
```prisma

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model UserEntity {
  id            Int                 @id @default(autoincrement())
  email         String              @unique
  firstName     String
  lastName      String
  emailVerified Boolean?             @default(false)
  mobile        String?
}

```
**Generated Class** 
```ts
/**
 * Model UserEntity
 * 
 */
export type UserEntity = {
  id: number
  email: string
  firstName: string
  lastName: string
  emailVerified: boolean|null
  mobile: string | null

}

```

### Environment & setup

- OS: macOS,
- Database: SQLite
- Node.js version: v18.14.2

### Prisma Version

```
prisma                  : 4.11.0
@prisma/client          : 4.11.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/migration-engine-darwin)
Format Wasm             : @prisma/prisma-fmt-wasm 4.11.0-57.8fde8fef4033376662cad983758335009d522acb
Default Engines Hash    : 8fde8fef4033376662cad983758335009d522acb
Studio                  : 0.483.0
```

**Comment by nurul3101** [comment]
Hello @jega-ms 👋 

The generated `UserEntity` type has null types in the `emailVerified` and `mobile` fields, which would allow them to be Optional.
Can you please elaborate on what would you expect the generated type to look like?

**Reply by richardryangarcia** [reply]
Hi Im having this same issue. I think what the OP is saying is that the id type generated is not optional even though it has a default value in the schema file. Im getting the same when generating using prisma client. When I go to create a record, I dont want to pass an id so that the default would be used but the transaction fails because I'm not passing an id 🙃 

Any help would be greatly appreciated !

**Reply by nurul3101** [reply]
Hello @richardryangarcia 👋 

The fields that are marked with the `@id` attribute are required and cannot be marked as optional with (?)

If you would remove the `@id` filter, the generated type would look like this:

<img width="864" alt="Screenshot 2023-03-24 at 2 51 20 PM" src="https://user-images.githubusercontent.com/13049676/227478858-b5b53615-12cc-4ea6-873b-9319a314f72a.png">

The id would be optional, but in this case, email would be required to determine a record uniquely.

Can you share an example query were not passing an id makes the transaction fail?

**Reply by jega-jsan** [reply]
@richardryangarcia I'm also facing the similar issue for ID, My current workaround is typecasting. it solves the ID issue. 

```
     {  email: source.email,
            firstName: source.firstName,
            lastName: source.lastName,
            emailVerified: false,
            roleId: 1,
        }  as UserEntity;

```

**Comment by jega-jsan** [comment]
hi @nurul3101 ,  sorry for the late response.  my actually requirement was as @richardryangarcia explained in above code. prisma should generate class something like below for optional fields, so that we don't have to pass the ID value or null check for optional fields. 

```
export type UserEntity = {
  id: number
  email: string
  firstName: string
  lastName: string
  emailVerified?: boolean
  mobile?: string 
}
```

**Reply by nurul3101** [reply]
So essentially if a default attribute is provided (Example: `@default(autoincrement())` in `id` field), then the generated type should have that field as optional, correct?

**Reply by nurul3101** [reply]
Can you maybe even share the query that you are invoking in which you have to pass the `id`?

**Reply by jega-jsan** [reply]
no,  When we define optional fields in schema e.g mobile, generator should create model class as mentioned below. 

e.g 

```
model UserEntity {
  id            Int                 @id @default(autoincrement())
  mobile        String?
}

```

Model class : 

```
export type UserEntity = {
  mobile?: string 
}
```

**Reply by ahmadawji** [reply]
Hey, please refer to this example: https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types

You can customize the generated model type using Prisma Client's helper types. Model types only contain the model's [scalar](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#scalar-fields) fields, but doesn't account for any relations. That's because [relations are not included by default](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#return-the-default-selection-set) in Prisma Client queries.
However, sometimes it's useful to have a type available that includes a relation (i.e. a type that you'd get from an API call that uses [include](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields#include-relations-and-select-relation-fields)).
**Example:**
The following example uses the Prisma.validator to create two type-safe objects and then uses the Prisma.UserGetPayload utility function to create a type that can be used to return all users and their posts.
```
import { Prisma } from '@prisma/client'

// 1: Define a type that includes the relation to `Post`
const userWithPosts = Prisma.validator<Prisma.UserDefaultArgs>()({
  include: { posts: true },
})

// 2: Define a type that only contains a subset of the scalar fields
const userPersonalData = Prisma.validator<Prisma.UserDefaultArgs>()({
  select: { email: true, name: true },
})

// 3: This type will include a user and all their posts
type UserWithPosts = Prisma.UserGetPayload<typeof userWithPosts>
```

**Comment by richardryangarcia** [comment, marked as answer]
So in my situation, I realized that `prisma db pull` pulls the prisma.schema but it DOES NOT include the defaults. I manually added the default and regenerated the client, then it had the expected types (optional id). Not sure if that is working as intended but db pull ideally would pull an exact copy of the schema.prisma file including any defaults.

**Reply by janpio** [reply]
It should. If it does not, that means your database is either not in the state of your previous `schema.prisma` file or you hit a bug. Please open a bug report issue if that is the case.

**Reply by richardryangarcia** [reply]
I also noticed that when defining unique keys on a table, db pull also does not include the name of the field thats defined in the main prisma.schema file. 
I will open a bug report to get these looked at. If anyone else is not seeing this issue please let me know.
